### DISCUSSION TITLE
testing dpi changes

### DISCUSSION THREAD

**Discussion by trebahl** [dotnet/wpf#10090] [General]
I would like to write automated tests at various dpi settings; also I would like to change the dpi during the test. Changing the dpi of the machine running the tests is cumbersome and fragile, so I would prefer to achieve the same effect inside the test process. Is there a way?

I was thinking of using reflection to call HwndSource.ChangeDpi. Is that a viable solution? Or does something in wpf's guts call native functions that will still be reporting the real dpi?

Which value should I pass as `suggestedRect` to the constructor of HwndDpiChangedEventArgs ?

**Comment by miloush** [comment]
That looks like a reasonable thing to try. You can pass `Rect.Empty` for the suggestedRect.

**Reply by miloush** [reply]
Technically you could send WM_DPICHANGED to the window if that sounded better than reflection.

**Comment by trebahl** [comment]
Thanks. With Rect.Empty, the window disappears from the screen and the code gets stuck in an infinite loop. I tried the following instead, which seems to work:

```
            var before = VisualTreeHelper.GetDpi(w);
            var pos = w.PointToScreen(new Point(0, 0));
            var sz = w.PointToScreen(new Point(w.ActualWidth, w.ActualHeight)) - pos;
            var mul = scale / before.DpiScaleY;
            sz = mul * sz;
            var r = new Rect(pos, new Size(Math.Ceiling(sz.X), Math.Ceiling(sz.Y))); // used in HwndDpiChangedEventArgs
```

**Reply by miloush** [reply]
Interesting, the code explicitly states that should be valid:

https://github.com/dotnet/wpf/blob/d73dd1dbd908398cef7f50e57812e55b90ed3676/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/DpiChangedEventArgs.cs#L63-L69

Oh well, but anyway yes if you want the window to resize too, you do need to provide the new Rect for it apparently.

**Reply by trebahl** [reply]
I got confused, it's not an infinite loop, it's just the normal event loop waiting for an event. It seems the window disappears because its size becomes (0,0).

HwndSource.OnDpiChanged calls HwndTarget.OnDpiChanged which calls UnsafeNativeMethods.SetWindowPos with the supplied rectangle, without checking if it is empty.
